qa: qa_filterbank finds some aligment issues

```
269/360 Test #269: qa_filterbank ......................................***Failed    0.57 sec
gr::buffer::allocate_buffer: warning: tried to allocate
   8 items of size 8000. Due to alignment requirements
   64 were allocated.  If this isn't OK, consider padding
   your structure to a power-of-two bytes.
   On this platform, our allocation granularity is 4096 bytes.
gr::buffer::allocate_buffer: warning: tried to allocate
   8 items of size 8000. Due to alignment requirements
   64 were allocated.  If this isn't OK, consider padding
   your structure to a power-of-two bytes.
   On this platform, our allocation granularity is 4096 bytes.
F
======================================================================
FAIL: test_000 (__main__.test_filterbank_vcvcf)
From: https://ci.gnuradio.org/buildbot/#/builders/49/builds/331/steps/5/logs/stdio
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/src/pull/2204/master/gr-filter/python/filter/qa_filterbank.py", line 128, in test_000
    self.assertComplexTuplesAlmostEqual(comb_results2, outdata2, 6)
  File "/data/src/pull/2204/master/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 86, in assertComplexTuplesAlmostEqual
    self.assertComplexAlmostEqual (a[i], b[i], places, msg)
  File "/data/src/pull/2204/master/gnuradio-runtime/python/gnuradio/gr_unittest.py", line 54, in assertComplexAlmostEqual
    msg or '%r != %r within %r places' % (first, second, places))
AssertionError: (0.03780283836319677-0.13573173385334286j) != (0.04380534589290619+0.08077553659677505j) within 6 places
----------------------------------------------------------------------
Ran 1 test in 0.047s
FAILED (failures=1)
```

I'm pulling this out of #1261 since I couldn't reproduce this on my on machine.
Bash line to test for failures:
```
for i in {1..1000}; do echo "$i. run"; result=$(ctest --output-on-failure -R qa_filterbank 2>&1); if [ $? -ne 0 ]; then echo "$i. run failed\n"; echo $result >> /tmp/error.log; fi; done
```
It's not entirely clear if that's a CI shortcoming or a serious issue.

--- a/gr-blocks/lib/CMakeLists.txt
+++ b/gr-blocks/lib/CMakeLists.txt
@@ -222,6 +222,10 @@ IF(MSVC)
     )
 ENDIF(MSVC)
 
+if(MSVC)
+    target_compile_definitions(gnuradio-blocks PUBLIC _ENABLE_ATOMIC_ALIGNMENT_FIX)
+endif()
+
 if(BUILD_SHARED_LIBS)
     if (SNDFILE_FOUND)
         GR_LIBRARY_FOO(gnuradio-blocks SNDFILE)

--- a/gr-blocks/lib/probe_signal_impl.h
+++ b/gr-blocks/lib/probe_signal_impl.h
@@ -13,6 +13,7 @@
 #define PROBE_SIGNAL_IMPL_H
 
 #include <gnuradio/blocks/probe_signal.h>
+#include <atomic>
 
 namespace gr {
 namespace blocks {
@@ -21,7 +22,7 @@ template <class T>
 class probe_signal_impl : public probe_signal<T>
 {
 private:
-    T d_level;
+    std::atomic<T> d_level;
 
 public:
     probe_signal_impl();

--- a/gr-blocks/lib/probe_signal_v_impl.cc
+++ b/gr-blocks/lib/probe_signal_v_impl.cc
@@ -47,6 +47,7 @@ int probe_signal_v_impl<T>::work(int noutput_items,
 {
     const T* in = (const T*)input_items[0];
 
+    gr::thread::scoped_lock guard(d_mutex);
     for (size_t i = 0; i < d_size; i++)
         d_level[i] = in[(noutput_items - 1) * d_size + i];
 
@@ -57,6 +58,13 @@ template class probe_signal_v<std::int16_t>;
 template class probe_signal_v<std::int32_t>;
 template class probe_signal_v<float>;
 template class probe_signal_v<gr_complex>;
+
+template <class T>
+std::vector<T> probe_signal_v_impl<T>::level() const
+{
+    gr::thread::scoped_lock guard(d_mutex);
+    return d_level;
+}
 
 } /* namespace blocks */
 } /* namespace gr */

--- a/gr-blocks/lib/probe_signal_v_impl.h
+++ b/gr-blocks/lib/probe_signal_v_impl.h
@@ -23,12 +23,13 @@ class probe_signal_v_impl : public probe_signal_v<T>
 private:
     std::vector<T> d_level;
     const size_t d_size;
+    mutable gr::thread::mutex d_mutex;
 
 public:
     probe_signal_v_impl(size_t size);
     ~probe_signal_v_impl() override;
 
-    std::vector<T> level() const override { return d_level; }
+    std::vector<T> level() const override;
 
     int work(int noutput_items,
              gr_vector_const_void_star& input_items,

--- a/gr-blocks/python/blocks/qa_probe_signal.py
+++ b/gr-blocks/python/blocks/qa_probe_signal.py
@@ -49,6 +49,45 @@ class test_probe_signal(gr_unittest.TestCase):
         self.assertEqual(len(output), vector_length)
         self.assertAlmostEqual(value[3], output[3], places=6)
 
+    def test_003_race_condition_regression_test(self):
+        src = blocks.vector_source_c([1 + 2j, 3 + 4j], True)
+        dst = blocks.probe_signal_c()
+
+        self.tb.connect(src, dst)
+        self.tb.start()
+
+        while dst.level() == 0.0:
+            continue
+
+        for _ in range(100000):
+            output = dst.level()
+            self.assertIn(output, [1 + 2j, 3 + 4j])
+
+        self.tb.stop()
+        self.tb.wait()
+
+    def test_004_race_condition_regression_test_vector(self):
+        vector_length = 10
+        src_data = [1.0] * vector_length + [2.0] * vector_length
+
+        src = blocks.vector_source_f(src_data, True, vector_length)
+        dst = blocks.probe_signal_vf(vector_length)
+
+        self.tb.connect(src, dst)
+        self.tb.start()
+
+        while dst.level()[0] == 0.0:
+            continue
+
+        for _ in range(10000):
+            output = dst.level()
+            self.assertIn(output[0], [1.0, 2.0])
+            for i in range(1, vector_length):
+                self.assertEqual(output[0], output[i])
+
+        self.tb.stop()
+        self.tb.wait()
+
 
 if __name__ == '__main__':
     gr_unittest.run(test_probe_signal)

